pkg/apis/kubedirector/v1beta1/kubedirectorcluster_types.go

```diff
@@ -25,11 +25,20 @@ import (
 // requested cluster roles, each of which will be implemented (by KubeDirector)
 // using a StatefulSet.
 type KubeDirectorClusterSpec struct {
-	AppID         string    `json:"app"`
-	AppCatalog    *string   `json:"appCatalog,omitempty"`
-	ServiceType   *string   `json:"serviceType,omitempty"`
-	Roles         []Role    `json:"roles"`
-	DefaultSecret *KDSecret `json:"defaultSecret,omitempty"`
+	AppID               string      `json:"app"`
+	AppCatalog          *string     `json:"appCatalog"`
+	ServiceType         *string     `json:"serviceType"`
+	Roles               []Role      `json:"roles"`
+	DefaultSecret       *KDSecret   `json:"defaultSecret"`
+	ConfigMetaGenerator int32       `json:"configMetaGenerator"`
```

Comment:
"configmeta" is kind of a confusing inside-baseball term for the JSON file, made a little more confusing by there also being a configMeta dictionary property inside the KD App.

IIRC this property, and the associated status property below, are intended to track changes specifically to the connected ConfigMaps. If that's the case then let's reflect that in the property name, e.g. we could call this one configMapGeneration and the status property could be lastConfigMapGeneration.

Also, can we store these as int64 rather than int32?

Reply:
Hmm I see this is also being used for reaction to kdcluster connections changes... I'll look over that in a bit.

Reply:
OK I see what you're doing there.

For discussion's sake:

There's a slightly more expensive way to handle this change detection that avoids the ugliness of adding something to the spec (which ideally only the enduser client would modify) and also avoids having to store all the connections info in the status stanza. Something I saw mentioned in a talk at KubeCon.

We could have a property in the status that is a hash of the connections. You can use DeepHashObject from k8s.io/kubernetes/pkg/util/hash ... probably other methods. So the validator wouldn't need to write anything to the spec, but every time through the reconciler the reconciler would need to do this hash and compare it to what is stored in the status.

I'll think further & sleep on this, and also on the question of whether the clusters/configmaps should be wrapped in a Connections object together or not. (You too!)

Reply:
Thinking now about the case where some cluster-I'm-connected to changes; if we were doing the hash approach, what would trigger me to update my connections? Currently you have the reconciler for the changed cluster doing a write back to my spec.

In the hash approach, we'd have to include something from each cluster-we're-connected to in the hash. I think we'd only need to include the specGenerationToProcess (from the status stanza). Granted we'd have to read each connected kdcluster object to find it out, but that shouldn't be a LOT of clusters.

I know in your current code you also are trying to take care to only propagate changes from connected clusters once the connected cluster reaches a ready status, I think there would be ways to deal with that in this approach but I will hold off for now on going further down this rabbit hole.